### DISCUSSION TITLE
ControlFlow: Skipping activities and handling errors

### DISCUSSION THREAD

**Discussion by bauermalzwei** [elsa-workflows/elsa-core#4141] [Q&A]
Hi,

thanks for your work with this project. I'm currently working with ELSA 2.9 and I have trouble understanding how to build my workflows properly.
Maybe you can shed some light on what the underlying concept is that I might be missing:
Consider a very linear workflow with a few activities that are executed linearly:

```
public void Build(IWorkflowBuilder builder)
{
	builder
		.Then<FirstActivity>()
		  .WithName(nameof(FirstActivity))
		  .ThenNamed(nameof(SecondActivity))
		.Add<SecondActivity>()
			.WithName(nameof(SecondActivity))
			.ThenNamed(nameof(ThirdActivcity))
		.Add<ThirdActivity>()
			.WithName(nameof(ThirdActivity))
			.ThenNamed(nameof(FinalActivcity))
		.Add<FinalActivcity>()
			.WithName((nameof(FinalActivcity)))
			.WithDisplayName(nameof(FinalActivcity));
}
```
My question is, how do I properly setup some sort of error handling activity that is executed in case an activity returns errors?
If an activity returns Fault, e.g.
```
try
{
   DoSomethingFaulty();
   return Done();
}
catch (Exception e)
{
    Logger.Warn(e);
    return Fault(e);
}
```
I have found no way to actually hook up an activity in that case. Why would I need that? Imagine a cleanup that needs to happen as the last step of a workflow - regardless if all steps were successful/executed. So if an error occurs early in the workflow, it makes sense to skip ahead to the cleanup part.

I've played around with forking but that doesn't feel very effective and also forces me to use an explicit Outcome - which also seems clunky.

```
public void Build(IWorkflowBuilder builder)
    {
        builder
            .Then<FirstActivity>(fork =>
            {
                fork
                    .When("Done")
                    .ThenNamed(nameof(SecondActivity));
                fork
                    .When("Error")
                    .ThenNamed(nameof(CleanupActivity));
            }).Add<SecondActivity>()
                .WithName(nameof(SecondActivity))
                .ThenNamed(nameof(FinalActivcity))
            .Add<ThirdActivity>(branch: 
                fork =>
                {
                    fork
			.When("Done")
			.ThenNamed(nameof(FinalActivcity));
                    fork
			.When("Error")
			.ThenNamed(nameof(CleanupActivity));
                })
                .WithName(nameof(ThirdActivity))
            .Add<FourthActivity>()
                .WithName(nameof(FourthActivity))
                .ThenNamed(nameof(CleanupActivity))
            .Add<CleanupActivity>()
                .WithName(nameof(CleanupActivity))
                .ThenNamed(nameof(FinalActivcity))
            .Add<FinalActivcity>()
                .WithName((nameof(FinalActivcity)))
                .WithDisplayName(nameof(FinalActivcity));
    }
```
I'm assuming I am missing some concepts here on how to approach this. I would be very happy if someone could point me in the right direction.

Thank you!

**Comment by sfmskywalker** [comment]
Hello @bauermalzwei,

Thanks for highlighting this. Handling errors and ensuring proper cleanup is essential. In Elsa 2, you have two primary ways to manage this:

**1. Error Outcome**

Instead of returning a Fault in custom activities, you can return an "Error" outcome. This lets you directly attach logic to this outcome in your workflow. For instance:

```csharp
try
{
   DoSomethingFaulty();
   return Done();
}
catch (Exception e)
{
    Logger.Warn(e);
    return Outcome("Error");
}
```

You can then implement it in your workflow as shown in your example. This is a straightforward approach and aligns with your intuition.

**2. Compensable Activity**

Elsa 2 introduces the `Compensable` activity. It allows you to define steps to execute if a fault occurs, essentially offering a "cleanup" branch named "Compensate". 

Using `Compensable`, your workflow can be structured as:

```csharp
public void Build(IWorkflowBuilder builder)
{
    builder
        .Then<Compensable>(compensable =>
        {
            compendable
                .When("Compensate") // This outcome is scheduled when any of the activities after Compensable faults.
                .Then<CleanupActivity>();
        })
        .Then<FirstActivity>()
        .Then<SecondActivity>()
        .Then<ThirdActivity>()
        .Then<FinalActivcity>();
}
```

For a hands-on example, check [this sample workflow](https://github.com/elsa-workflows/elsa-core/blob/master/src/samples/console/Elsa.Samples.CompensationConsole/Workflows/CompensableWorkflow.cs).

I hope this clears things up. Let us know if you need further assistance!
